### PR TITLE
1528 refactor generation service

### PR DESCRIPTION
**Pull request by dgomezc**
## PR checklist

- Quick summary of changes
   - Change userSelection service to services array
   - Refactor generation module
   - Add generationService class
   - Add differents generation classes to generate differents services
- Which issue does this PR relate to?
Refactor generation module in extension #1528

### CODE REVIEW

File: src/extension/src/utils/generation/generators/TemplatesGenerator.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import { MESSAGES } from "../../../constants/messages";
+import { CoreTemplateStudio } from "../../../coreTemplateStudio";
+import { IGenerationData } from "../../../types/generationPayloadType";
+import {
+  GenerationItemStatus,
+  sendGenerationStatus,
+  TEMPLATES_GENERATION_NAME,
+  updateStatusMessage,
+} from "../../generationStatus";
```
Comment (prupipho) [inline src/extension/src/utils/generation/generators/TemplatesGenerator.ts]: I think **_generationStatus_** file makes sense within de **_generation_** folder. Currently one level up under **_utils_**

File: src/extension/src/client-modules/generation.ts [hunk 2]
```diff
@@ -1,25 +1,17 @@
...
-import { Logger } from "../utils/logger";
-import { IGenerationPayloadType, IServicesGenerationPayload } from "../types/generationPayloadType";
-import { sendToClientGenerationStatus, GenerationItemStatus, updateStatusMessage, GENERATION_NAMES } from "../utils/generationStatus";
-import { MESSAGES } from "../constants/messages";
+  IAppService,
+  ICosmosDB,
+  IGenerationData,
+  SERVICE_CATEGORY,
+  SERVICE_TYPE,
+} from "../types/generationPayloadType";
```
Comment (prupipho) [inline src/extension/src/client-modules/generation.ts]: Is _**generationPayloadType**_ still a valid file name or maybe just _**generationType**_ could work?

File: src/extension/src/client-modules/generation.ts [hunk 3]
```diff
@@ -221,22 +48,49 @@ export class Generation extends WizardServant {
...
+        resourceGroup,
+        location,
+        serviceName: accountName,
+        api,
+      };
+      generationData.services?.push(cosmosDB);
+    }
 
-  //TODO: Move to telemetryService?
-  private deployWithTelemetry<T>(telemetryEvent: string, callback: Promise<T>): Promise<any> {
```
Comment (prupipho) [inline src/extension/src/client-modules/generation.ts:235]: this function _seems_ duplicated both in GenerationService & ResourceGroupGenerator. Could it make sense make it public and have it just in one place? Not sure where it fits most to be honest... I´ll think about it..

File: src/extension/src/utils/generation/GenerationService.ts [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+  public rejectServices(services: Array<IService>) {
+    const { SERVICE_DEPLOYMENT_HALTED } = MESSAGES.GENERATION;
+    services.forEach((service) => {
+      const generator = this.serviceGenerators.get(service.type);
+      if (generator)
+        sendGenerationStatus(generator.serviceType, GenerationItemStatus.Failed, SERVICE_DEPLOYMENT_HALTED);
+    });
+  }
+
+  public async generateServices(generationData: IGenerationData) {
```
Comment (prupipho) [inline src/extension/src/utils/generation/GenerationService.ts]: I think this one can be a _**private**_ method

File: src/extension/src/types/generationPayloadType.ts [hunk 5]
```diff
@@ -1,34 +1,55 @@
...
-export interface ICosmosDBGenerationPayload {
+export interface IAzureService extends IBaseService{
+  category: SERVICE_CATEGORY.AZURE;
   subscription: string;
   resourceGroup: string;
   location: string;
-  accountName: string;
+}
+
+export interface IAppService extends IAzureService {
```
Comment (prupipho) [inline src/extension/src/types/generationPayloadType.ts]: Not sure but could it make sense to move those interfaces and enums to somewhere else? Or maybe it´s just the name of the file generationPayloadType. The only type that seems tied to "generation" it´s just generationData, right? The rest sounds more generic just related to Services... Maybe have a new file _**serviceType**_ ?

File: src/extension/src/utils/generation/GenerationService.ts [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+        this.addToDeployQueue(generator.telemetryEventName, generator.generate(service, generationData));
+      }
+    });
+  }
+
+  private addToDeployQueue(telemetryEventName: string, callback: Promise<DeployedServiceStatus>) {
+    this.servicesQueue.push(this.deployWithTelemetry(telemetryEventName, callback));
+  }
+
+  private deployWithTelemetry<T>(telemetryEvent: string, callback: Promise<T>): Promise<any> {
```
Comment (prupipho) [inline src/extension/src/utils/generation/GenerationService.ts:99]: this function seems duplicated both in GenerationService & ResourceGroupGenerator. Could it make sense make it public and have it just in one place? Not sure where it fits most to be honest... I´ll think about it..

File: src/extension/src/utils/generation/generators/ResourceGroupGenerator.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+  private async deployResourceGroup(resourceGroup: ResourceGroupSelection): Promise<void> {
+    try {
+      await AzureServices.deployResourceGroup(resourceGroup);
+    } catch (error) {
+      Logger.appendError("EXTENSION", MESSAGES.ERRORS.CREATING_RESOURCE_GROUP, error);
+    }
+  }
+
+  private deployWithTelemetry<T>(telemetryEvent: string, callback: Promise<T>): Promise<any> {
```
Comment (prupipho) [inline src/extension/src/utils/generation/generators/ResourceGroupGenerator.ts:36]: this function seems duplicated both in GenerationService & ResourceGroupGenerator. Could it make sense make it public and have it just in one place? Not sure where it fits most to be honest... I´ll think about it..

Comment (prupipho) [reply src/extension/src/utils/generation/generators/ResourceGroupGenerator.ts:36]: At the moment this is the best place to be. Duplicated. Waiting for a future refactor.

**Review by sibille — APPROVED** [review]

**Pull request merged**